build_resolvers/lib/src/analysis_driver_model.dart

```diff
@@ -19,36 +25,148 @@ import 'package:build/build.dart';
       List<AssetId> entryPoints,
       Future<void> Function(
               FutureOr<void> Function(AnalysisDriverForPackageBuild))
           withDriverResource,
-      {required bool transitive});
+      {required bool transitive}) async {
+    /// TODO(davidmorgan): add test coverage for whether transitive
+    /// sources are read when [transitive] is false, fix the implementation
+    /// here.
+    /// TODO(davidmorgan): add test coverage for whether
+    /// `.transitive_deps` files cut off the reporting of deps to the
+    /// [buildStep], fix the implementation here.
+
+    // Find transitive deps, this also informs [buildStep] of all inputs).
+    final ids = await _expandToTransitive(buildStep, entryPoints);
+
+    // Apply changes to in-memory filesystem.
+    for (final id in ids) {
+      if (await buildStep.canRead(id)) {
+        final content = await buildStep.readAsString(id);
+
+        /// TODO(davidmorgan): add test coverage for when a file is
+        /// modified rather than added, fix the implementation here.
+        resourceProvider.newFile(id.asPath, content);
+      } else {
+        if (resourceProvider.getFile(id.asPath).exists) {
+          resourceProvider.deleteFile(id.asPath);
+        }
+      }
+    }
+
+    // Notify the analyzer of changes.
+    await withDriverResource((driver) async {
+      for (final id in ids) {
+        // TODO(davidmorgan): add test coverage for over-notification of
+        // changes, fix the implementaion here.
+        driver.changeFile(id.asPath);
+      }
+      await driver.applyPendingFileChanges();
+    });
+  }
+
+  /// Walks the import graph from [ids], returns full transitive deps.s
+  Future<Set<AssetId>> _expandToTransitive(
+      AssetReader reader, Iterable<AssetId> ids) async {
+    final result = <AssetId>{};
+    await __expandToTransitive(reader, ids, result);
+    return result;
+  }
+
+  /// Walks the import graph from [ids], ignoring nodes already in [result].
+  ///
+  /// Call with [result] empty to add full transitive deps to [result].
+  Future<void> __expandToTransitive(
+      AssetReader reader, Iterable<AssetId> ids, Set<AssetId> result) async {
+    final nextIds = Queue.of(ids);
+    while (nextIds.isNotEmpty) {
+      final nextId = nextIds.removeFirst();
+
+      // Skip if already seen.
+      if (!result.add(nextId)) continue;
+
+      // Skip if not readable.
+      if (!await reader.canRead(nextId)) continue;
+
+      final content = await reader.readAsString(nextId);
+      final deps = _parseDependencies(content, nextId);
+      nextIds.addAll(deps.where((id) => !result.contains(id)));
+    }
+  }
+}
+
+const _ignoredSchemes = ['dart', 'dart-ext'];
+
+/// Parses Dart source in [content], returns all depedencies: all assets
+/// mentioned in directives, excluding `dart:` and `dart-ext` schemes.
+List<AssetId> _parseDependencies(String content, AssetId from) =>
+    parseString(content: content, throwIfDiagnostics: false)
```

Comment:
Stuff like this is completely unreadable to me.

Reply:
Noted, thanks. This is direct from the current implementation, so I'll leave as-is for now.